torch_xla/csrc/runtime/ifrt_computation_client.h

```diff
@@ -0,0 +1,222 @@
+      absl::Span<const std::string> devices,
+      const ExecuteReplicatedOptions& options) override;
+
+  size_t GetNumDevices() const override;
+
+  std::string GetDefaultDevice() const override;
+
+  std::vector<std::string> GetLocalDevices() const override;
+
+  std::vector<std::string> GetAllDevices() const override;
+
+  int GetProcessIndex() const override { return client_->process_index(); };
+
+  int GetNumProcesses() const override;
+
+  const absl::flat_hash_map<
+      std::string, torch_xla::runtime::ComputationClient::DeviceAttribute>&
+  GetDeviceAttributes(const std::string& device) override;
+
+  void SetReplicationDevices(
+      std::shared_ptr<std::vector<std::string>> devices) override;
+
+  std::shared_ptr<std::vector<std::string>> GetReplicationDevices() override;
+
+  void WaitDeviceOps(absl::Span<const std::string> devices) override;
+
+  std::map<std::string, Metric> GetMetrics() const override;
+
+  void InitializeCoordinator(int global_rank, int world_size,
+                             std::string master_addr,
+                             std::string port) override;
+
+  XlaCoordinator& GetCoordinator() override;
+
+  bool CoordinatorInitialized() const override;
+
+  torch::lazy::hash_t HashCompilationEnv() override { return comp_env_hash_; }
+
+  // NOT IMPLEMENTED
+
+  MemoryInfo GetMemoryInfo(const std::string& device) override {
+    XLA_ERROR() << __FUNCTION__ << " not implemented";
+  };
+
+  std::string SerializeComputation(const ComputationPtr computation) override {
+    XLA_ERROR() << __FUNCTION__ << " not implemented";
+  }
+
+  ComputationPtr DeserializeComputation(
+      const std::string& serialized) override {
+    XLA_ERROR() << __FUNCTION__ << " not implemented";
+  }
+
+ private:
+  std::shared_ptr<xla::ifrt::PjRtClient> client_;
+  std::unique_ptr<XlaCoordinator> coordinator_;
+  // global_ordinals_ tracks a map from PjRtDeviceId to the device's
+  // dense global ordinal.
+  std::unordered_map<int, int> global_ordinals_;
+  std::unordered_map<std::string, xla::PjRtDevice* const> string_to_device_;
+  std::shared_ptr<std::vector<std::string>> replication_devices_;
+  OperationManager operation_manager_;
+  tsl::thread::ThreadPool pool_ = tsl::thread::ThreadPool(
+      tsl::Env::Default(), "ifrt", std::thread::hardware_concurrency());
+  torch::lazy::hash_t comp_env_hash_;
+
+  xla::ifrt::Device* StringToIfrtDevice(const std::string& device);
+
+  std::string IfrtDeviceToString(xla::ifrt::Device* const device) const;
+  std::vector<std::string> IfrtDevicesToString(
+      absl::Span<xla::ifrt::Device* const> devices) const;
+
+  struct IfrtData : public Data {
+    IfrtData(std::string device, xla::Shape device_shape)
+        : Data(std::move(device), std::move(device_shape)) {}
+
+    IfrtData(std::string device, xla::Shape device_shape,
+             tsl::RCReference<xla::ifrt::Array> buffer,
+             std::optional<xla::OpSharding> sharding = std::nullopt)
```

Comment:
I wonder if IfrtData needs to track the sharding, or can we just pull it from the underlying Array? Same for device_shape.

I assume there may be some overhead reading these off of the array, so it may be better to track these separately.

Reply:
> I wonder if IfrtData needs to track the sharding, or can we just pull it from the underlying Array? Same for device_shape.

So, the IFRT array does carry its own sharding, but it's in IFRT's format. There is an IFRT wrapper over `xla::HloSharding` (aka the non-proto version of `xla::OpSharding`), but I was running into crashes while trying to use it on our input data. Also, IFRT does not guarantee that the arrays returned after execution specifically have the `HloSharding` type.

This is a place where the IFRT abstraction doesn't really work for us, since we rely on the `xla::OpSharding`/`xla::HloSharding` for building HLOs. To use `xla::ifrt::Array`'s sharding full-time, we'd have to update our code to operate on `xla::ifrt::Sharding`. For now, I'm recycling most of our existing logic that we used for `PjRtShardedData`'s sharding.